Custom interface for controller button/axis state (Variant A)

toji opened the pull request:
This is a proposal to fix the long-running #392, as well as address #336. (And honestly those two could be separated if one part or the other of this proves controversial, but for now I think it's useful to see them together.)

The PR is mostly a more formalized version of the sketched out proposal in #392, with some of the edge cases thought through a bit more. If there was any particular piece of this that I'm not really comfortable with, it's the name 'XRTrackedControllerState`, because I don't feel like that really communicates what it is very well. The term that comes to mind for me immediately is "XRTrackedControllerInput", but that makes for a bit of a namespace conflict with things like "XRInputSource". Problem is that neither Nell or I have come up with a particularly good alternative. ([Twitter has opinions on this](https://twitter.com/Tojiro/status/1075463933119483904), FWIW. Most of which are "Input!" 😉)

So anyway, if you have a better name then please please please PLEASE share it! Any beyond that we'd love feedback on whether or not this feels like a system that will enable whatever particular input usecase you and your app/browser/platform/etc. have in mind.

toji commented:
Communicating a piece of offline feedback that @cwilso gave to me: The naming of the `.controller` attribute on the `XRInputSource` caused some confusion on his first readthrough because he felt it implied that the attributed represented the entirety of the controller device rather than communicating a subset of the input source's state (specifically the button/axis state). An attribute name along of the lines of (but less terrible than) `.buttonAxisState` rather than `.controller` would have better implied the right relationship between objects in his mind.

thetuvix commented on input-explainer.md:
```diff
@@ -280,6 +294,91 @@ When the canvas receives a `pointerdown` event an `XRInputSource` is created wit
+    }
+
+    // Determine if the controller has at least one extended input, such as a
+    // face button.
```
If we pursue an `extended` approach, we'll need to explicitly think through how buttons/axes will transition from `extended` to full named attributes.  Otherwise, promoting a button or axis will break existing content.

It may be more compatible to change `extended` to `elements` or such, including both the core and extended elements.  That way, there is no temptation for UAs to remove elements as they are promoted to core.

thetuvix replied:
From the call, it sounds like there are concerns about `element`, `component` or `part`.  However, we should agree on the structure we determine that we need here, and then find the least-bad name.

NellWaliczek replied:
I'm open to renaming `extended` to `all` or something similar with the appropriate behavior.  
That said, IIRC, webidl doesn't allow for a dictionary of values as an attribute.  At one point we considered a function like `getInputPart(string partName)`.  Perhaps that's an option?

thetuvix commented on input-explainer.md:
```diff
@@ -177,13 +177,27 @@ Most applications will want to visually represent the input sources somehow. The
+ - The `renderId` may be `'unknown'` if the type of input source cannot be reliably identified or the UA determines that the input source type must be masked for any reason.
+ - Otherwise the `renderId` should be a lower-case string that describes the physical input source. It should not include and indication of the handedness of the input source, as that should be implied by the `handedness` attribute.
+   - If the input source is a tracked hand without any held controller, the `renderId` MUST be `hand`
+   - For most controllers this should ideally be of the format `<vendor>-<product-id>`. For example: `oculus-touch`. UAs should make an effort to align on the strings that are returned for any given device.
```
> UAs should make an effort to align on the strings that are returned for any given device.

We should explicitly dig in on this at the F2F to work out some clever approach for going about this even without lots of cross-UA coordination before the first UA ships support for a device.

ddorwin replied:
https://www.w3.org/wiki/Repositories appears to be the latest thinking on such things and seems like a good fit.

I think this is a large enough topic that we should break the renderId part out of this PR so we can focus on controller state ergonomics and the logistics around this ID separately.

cwilso replied:
I'd caution against taking the thinking in that wiki as a current, direct solution.  It was a work item of the Advisory Board (while I was on it), hasn't been completed (or worked on in a while?) and it's not an adopted part of the W3C Process (translation: this hasn't been tested by any other repository/schema-like thing in W3C-space that I know of).

kearwood commented:
Some initial feedback:
- Value ranges might need to reflect values on both sides of "engaged".  Eg, some devices can be "more engaged".  Pressure sensitive buttons or grips could fall into this category.
- "name" of inputs perhaps should be split into multiple values
(One for localization, one as fixed string, One index identifier for skinned mesh models)
- If same device is available in Gamepad API, should it be referenced?

NellWaliczek commented:
@kearwood you mentioned
>If same device is available in Gamepad API, should it be referenced?

That shouldn't ever be the case.  This design is really just intended for tracked controllers, and it's intended they would never show up in the gamepads array.

NellWaliczek commented:
Also at the risk of saying something that's either shooting myself in the foot... or making @toji stare at me like -_-.... Looking at the `XRTrackedController` design right now.... it occurs to me that there really isn't anything tracked about it.  It's really just a nicer `Gamepad` attached an `XRInputSource`.  Which brings up two thoughts:
1. Should this idea of named buttons/trackpads and strongly typed data be proposed as a separate spec from WebXR?  Perhaps as a gamepad 2.0?
2. If that's the case, perhaps the right thing for the WebXR Level1 spec is to do what it was @kearwood  thought we were suggesting...and instead just have a `Gamepad` member.  That means the Gamepad spec could update independently to add the features described above and, so long as it's done in a way that can be feature detected, developers wouldn't be negatively impacted.

>note: I am totally NOT suggesting we keep the gamepad extensions.

*whispers* please don't hate me.....

NellWaliczek commented:
Oh, and one small aside... a colleague of mine noted that the `grip` and the `gripSpace` are a bit confusing due to their similar names... My inclination would be to change the name of `gripSpace` to something more like `renderSpace`

thetuvix commented:
> If that's the case, perhaps the right thing for the WebXR Level1 spec is to do what it was @kearwood  thought we were suggesting...and instead just have a `Gamepad` member.  That means the Gamepad spec could update independently to add the features described above and, so long as it's done in a way that can be feature detected, developers wouldn't be negatively impacted.

This is interesting to explore, given where we've landed with input source spaces.  The key thing we'd have to make sure is that whatever Gamepad 1.0 mapping we come up with avoids all the pitfalls we hit with WebVR's use of the Gamepad API:
* Well-known elements like "menu button" and "touchpad X" would need well-known button and axis indices in some kind of `"xr"` Gamepad `mapping`.  UAs would need to support gaps in both lists when a controller is missing a given element.
* For future extensibility of the core mapping as we standardize additional elements, we could have `"xr2"`/`"xr3"` or such.  This allows an app still using `"xr"` to get the old mapping, where non-standard buttons for that controller show up with their old non-standard indices.
* We'll have to work out some way to represent touch states for now - we could fall back to treating each button's touch amount as an additional sibling "button", but that'd be unfortunate.  Maybe it's a small enough tweak to add `touched` next to `pressed` on `GamepadButton`.

daoshengmu commented:
> 
> interface XRTrackedController {
>   readonly attribute XRTrackedControllerStateDouble? trigger;
>   readonly attribute XRTrackedControllerStateVector2? joystick;
>   readonly attribute XRTrackedControllerStateVector2? touchpad;
>   readonly attribute XRTrackedControllerStateDouble? grip;
>   readonly attribute XRTrackedControllerStateDouble? menu;
>   readonly attribute XRTrackedControllerStateDouble? a;
>   readonly attribute XRTrackedControllerStateDouble? b;
> };
> 
I love this way, at least, we don't need to care about the button and axis order as `Gamepad` API. However, we can't cover all of buttons in world. For example, Oculus Touch has x, y buttons. If we wanna add new buttons, we have to modify this `.webidl` and update it to the spec.

The other thing is I am thinking about handness inputs, like SteamVR Knuckles. They would expose hands as bone matrices, so Float32Array is an another type of `XRTrackedControllerState` that we should consider.

So, I was thinking to use action-base approach (https://github.com/immersive-web/webxr/issues/392#issuecomment-431200201), and Firefox has shipped OpenVR action-based input API in 65. My proposal of action-based like API is not really webby, and I am also not good at defining Web APIs, but this is just my two cents that might be helpful.

thetuvix commented:
> Oh, and one small aside... a colleague of mine noted that the `grip` and the `gripSpace` are a bit confusing due to their similar names... My inclination would be to change the name of `gripSpace` to something more like `renderSpace`

I've noticed a similar confusion - I often refer now to a handheld controller's **"palm pose"** rather than "grip pose" in other contexts to help reduce that confusion.  That would lead to `palmSpace`.

If we didn't want to presume a "palm" for generic renderable input sources, I'd rather just refer to the source's `rootSpace` rather than `renderSpace`, as I'd expect other source-relative data (e.g. Leap Motion hand joint positions) to make use of that same origin in the future.

toji commented:
Lots of feedback to sort through, but in the meantime I've rebased the issue now that the pose refactor has landed. I've also removed the `renderId` attribute (since that's now covered in #479).

toji commented on input-explainer.md:
```diff
@@ -280,6 +294,91 @@ When the canvas receives a `pointerdown` event an `XRInputSource` is created wit
+The UA may update the controller state at any point, which the restriction that it must remain constant during the running of each batch of `XRSession` `requestAnimationFrame` callbacks.
+
+> Note: You might wonder why `XRTrackedController` isn't tied to an `XRFrame` like the `XRInputPose` is. The answer is that data exposed by `XRFrame` is intended to be highly timing dependent, and have a limited useful lifespan. Poses are continuously changing, demand low latency, and are typically projected forward to a specific point in time. The also must be queried relative to a specific `XRF ...
+>
```
"must remain constant during Javascript execution" is basically the default assumption of Javascript, isn't it? Being single threaded, it would be highly unusual if values did change mid-execution without an obvious call to update them. rAF is special simply in that during the processing of rAF callbacks control is returned to the browser multiple times, once per callback, and yet we want the controller state to be consistent across all of them.

I'm looking at an alternative way of approaching this functionality, but I'll try to update all of this language to be clearer and take into account the rest of your feedback if it carries through to the next update.

dmarcos commented on input-explainer.md:
```diff
@@ -280,6 +280,91 @@ When the canvas receives a `pointerdown` event an `XRInputSource` is created wit
+
+Some applications need more than point-and-click style interaction provided by the `select` events. For devices that make use of controllers, more complete information about the state of those controller's inputs (buttons, triggers, touchpads, etc) can be observed via the `XRInputSource`'s `controller` attribute.
+
+The `controller` attribute is an instance of the `XRTrackedController` interface if the `XRInputState` represents an XR controller device with a subset of common inputs, or `null` otherwise. Examples of controllers that may expose their state this way include Oculus Touch, Vive wands, Oculus Go and Daydream controllers, or other similar devices. More traditional gamepads, such as XBox or Playstat ...
```
typo? discreet -> discrete

dmarcos replied:
Microsoft uses `Xbox` capitalization in [its materials](https://account.microsoft.com/account/xboxlive) vs. `XBox`

dmarcos replied:
[Sony uses](https://www.sony.com/electronics/playstation) `PlayStation` vs `Playstation`

toji commented:
This was discussed at the Jan 2019 F2F and pursuing a gamepad-based solution (as described in #499) was overwhelmingly preferred to this approach. Closing as a result in favor of devoting all input-related energy to PR #499.
